app/gui2/src/stores/visualization/metadata.ts

```diff
@@ -0,0 +1,34 @@
+import type { VisualizationModule } from '@/stores/visualization/compilerMessaging'
+import { ReactiveDb, ReactiveIndex } from '@/util/database/reactiveDb'
+import type { Opt } from '@/util/opt'
+import type { VisualizationIdentifier } from 'shared/yjsModel'
+
+export interface VisualizationMetadata extends Pick<VisualizationModule, 'name' | 'inputType'> {}
+
+function getTypesFromUnion(inputType: Opt<string>) {
+  return inputType?.split('|').map((type) => type.trim()) ?? ['Any']
+}
+
+declare const visualizationIdBrand: unique symbol
+export type VisualizationId = string & { [visualizationIdBrand]: never }
+
+export function toVisualizationId(meta: VisualizationIdentifier) {
+  return JSON.stringify({
+    // All fields MUST be explicitly written so that the order is consistent.
+    module: {
+      kind: meta.module.kind,
+      name: meta.module.kind === 'Library' ? meta.module.name : undefined,
+    },
+    name: meta.name,
+  }) as VisualizationId
+}
+
+export function fromVisualizationId(key: VisualizationId): VisualizationIdentifier {
+  return JSON.parse(key)
+}
+
+export class VisualizationMetadataDb extends ReactiveDb<VisualizationId, VisualizationMetadata> {
```

Comment:
On one hand I really don't like using `extends` and almost always prefer composition. But on the other hand, in this specific case it does look good, and maybe we should also move `SuggestionDb` to that same model. Will there ever be a situation where we want more `ReactiveDb` instances (i.e. tables) in the actual "database" type? If so, then we should really use composition. Otherwise I'm fine with this.

Reply:
hmm... fwiw i think it's fine to instead require `Db` classes to always be single tables - it's not like `ReactiveDb` supports multiple tables either, even though it is called a `Db`, right?

but yeah, the intent is for this to always be a single table - the name is supposed to mean, more or less, "a table associating each visualization type with its metadata"